coral-schema/src/main/java/com/linkedin/coral/schema/avro/SchemaUtilities.java

```diff
@@ -436,46 +445,62 @@ private static boolean isUnionSchemaCompatible(@Nonnull Schema leftSchema, @Nonn
       case STRING:
-      case NULL:
-        return leftSchema.getType() == rightSchema.getType();
+        outputSchema = leftSchema;
+        break;
       case FIXED:
-        boolean isSameType = leftSchema.getType() == rightSchema.getType();
-        boolean isSameNamespace = Objects.equals(leftSchema.getNamespace(), rightSchema.getNamespace());
-
-        return isSameType && (!strictMode || isSameNamespace);
+        compatible &= isSameNamespace(leftSchema, rightSchema, strictMode);
+        outputSchema = leftSchema;
+        break;
       case ENUM:
-        boolean isSameEnumType = (leftSchema.getType() == rightSchema.getType());
         boolean isSameSymbolSize = (leftSchema.getEnumSymbols().size() == rightSchema.getEnumSymbols().size());
-        boolean isSameEnumNamespace = Objects.equals(leftSchema.getNamespace(), rightSchema.getNamespace());
-
-        return isSameEnumType && isSameSymbolSize && (!strictMode || isSameEnumNamespace);
+        compatible &= isSameSymbolSize && isSameNamespace(leftSchema, rightSchema, strictMode);
+        outputSchema = leftSchema;
+        break;
       case RECORD:
-        return leftSchema.getType() == rightSchema.getType()
-            && isUnionRecordSchemaCompatible(leftSchema, rightSchema, strictMode);
-
+        outputSchema = mergeUnionRecordSchema(leftSchema, rightSchema, strictMode);
+        break;
       case MAP:
-        return leftSchema.getType() == rightSchema.getType()
-            && isUnionSchemaCompatible(leftSchema.getValueType(), rightSchema.getValueType(), strictMode);
-
+        Schema valueType = getUnionFieldSchema(leftSchema.getValueType(), rightSchema.getValueType(), strictMode);
+        outputSchema = Schema.createMap(valueType);
+        break;
       case ARRAY:
-        return leftSchema.getType() == rightSchema.getType()
-            && isUnionSchemaCompatible(leftSchema.getElementType(), rightSchema.getElementType(), strictMode);
-
+        Schema elementType = getUnionFieldSchema(leftSchema.getElementType(), rightSchema.getElementType(), strictMode);
+        outputSchema = Schema.createArray(elementType);
+        break;
       case UNION:
-        boolean isSameUnionType = (leftSchema.getType() == rightSchema.getType());
-        boolean isBothNullableType =
-            AvroSerdeUtils.isNullableType(leftSchema) && AvroSerdeUtils.isNullableType(rightSchema);
-
-        Schema leftOtherType = AvroSerdeUtils.getOtherTypeFromNullableType(leftSchema);
-        Schema rightOtherType = AvroSerdeUtils.getOtherTypeFromNullableType(rightSchema);
-        boolean isOtherTypeUnionCompatible = isUnionSchemaCompatible(leftOtherType, rightOtherType, strictMode);
-
-        return isSameUnionType && isBothNullableType && isOtherTypeUnionCompatible;
-
+        // If we are in this branch, we know that neither leftSchema nor rightSchema is a nullable union, but can still be a union
+        outputSchema = getUnionFieldSchema(leftSchema, rightSchema, strictMode);
```

Comment:
At present, this case doesn't happen because non-nullable union type like `[int, boolean]` is not supported in coral-hive.
But if it is supported in the future, I think 
`outputSchema = getUnionFieldSchema(leftSchema, rightSchema, strictMode);`
will cause infinite recursion.

Reply:
True. Removed.